_posts/blog/2018-01-24-pt-bug.md

```diff
@@ -0,0 +1,73 @@
+<!--more-->
+
+The M-Lab team is doing development to work around the bug in several ways:
+
+1. By changing how we call Paris Traceroute to reduce the probability that the bug condition occurs.
+2. By adding more sophisticated bad data detection and elimination code to our parsing of Paris Traceroute data.
+3. By adding monitoring to our parser so that we can know how much of our raw data is affected at any given time.
+4. By reprocessing our historical raw data archives in order to eliminate all bad traceroute data caused by this bug from our BigQuery database.
+
+Through disclosures and analyses like these, M-Lab re-confirms its commitment to open data and open science. Our raw data, warts, bugs, and all, is and will be always available [https://console.developers.google.com/storage/browser/m-lab/](https://console.developers.google.com/storage/browser/m-lab/), while our processed data [https://bigquery.cloud.google.com/dataset/measurement-lab:public?pli=1](https://bigquery.cloud.google.com/dataset/measurement-lab:public?pli=1) continues to reflect our best understanding of the state of the Internet represented by that raw data.
+
+## Read More: The bug and its fixes
+
+The root cause of the reported problem is a design flaw in Paris Traceroute, exacerbated by Rollins, [the wrapper script that M-Lab uses to run it](https://github.com/npad/sidestream/blob/master/paris_rollins.py). The Rollins wrapper script went into production in first half of 2016 (replacing an older script which had its own flaws).
+
+Every connection made to M-Lab systems triggers a Paris Traceroute to run from the M-Lab server back to the computer that initiated the connection. As the number of simultaneous traceroutes has increased, this has exposed a bug in Paris Traceroute where two independent traceroutes that overlap in time can become intermixed in the resulting Paris Traceroute output.  For example, if we were to simultaneously run a traceroute from server S to clients A and B, the S-A path reported by the tool might contain hops that are exclusive to the S-B path, or it could switch in the middle of Paris Traceroute's output from being the S-A path into being the S-B path.
+
+In the following example, the output from Paris Traceroute switches without warning from being the output for:
+
+`$ paris-traceroute 35.188.101.1`
+
+into containing the data for the contemporaneously-run command:
+
+`$ paris-traceroute 139.60.160.135`
+
+The resulting output from the first command (shown below) shows a link that doesn't exist from 216.239.51.185 to 173.239.28.18, on lines 8 and 9 and (worse!) switches without warning into the output from the second command (lines 9-13). We have highlighted the suspiciously large RTT at the end of line 9 (`**4802.776/4803.621/4807.622/1.790 ms**`):
+
+```bash
```

Comment:
Should be just the triple backtics. No `bash`.

Reply:
bash is required since the linter/formatter requires a language to properly highlight text in a fenced code block.